File: CORE/BACKEND/map_generator/graph_builder.py
import logging
import networkx as nx
from CORE.BACKEND.redis_tools import (
    save_to_redis, load_from_redis,
    KEY_RED_LINES, KEY_BLUE_CIRCLES, KEY_ADJACENCY,
    KEY_WHITE_LINES, KEY_GREEN_CIRCLES
)
from CORE.BACKEND.uid_utils import generate_uid, UIDPrefix
from .geometry_utils import haversine_distance

logger = logging.getLogger(__name__)

def identify_intersections():
    """Step 2: Blue Circles"""
    logger.info("GraphBuilder: Step 2 - Identifing Intersections")
    
    red_lines = []
    # Load from Redis (we utilize the side-effect of Step 1 saving to Redis)
    cached = load_from_redis(KEY_RED_LINES)
    if cached:
        for visual in cached:
            # Handle both old (list) and new (dict) formats
            coords = visual['path'] if isinstance(visual, dict) and 'path' in visual else visual
            if not isinstance(coords, list):
                continue

            for i in range(len(coords) - 1):
                p1 = (float(coords[i][0]), float(coords[i][1]))
                p2 = (float(coords[i+1][0]), float(coords[i+1][1]))
                red_lines.append((p1, p2))
    
    node_counts = {}
    adjacency = {}
    
    for start, end in red_lines:
        node_counts[start] = node_counts.get(start, 0) + 1
        node_counts[end] = node_counts.get(end, 0) + 1
        
        if start not in adjacency:
            adjacency[start] = set()
        if end not in adjacency:
            adjacency[end] = set()
        adjacency[start].add(end)
        adjacency[end].add(start)
        
    blue_circles = []
    relevant_nodes = set()

    # Create blue circles only for intersections/endpoints (count != 2)
    for node, count in node_counts.items():
        if count != 2:
            blue_circles.append({
                'id': generate_uid(UIDPrefix.BLUE_CIRCLE),
                'lat': node[0], 'lon': node[1],
                'connections': count
            })
            relevant_nodes.add(node)
    
    # Redis Save
    save_to_redis(KEY_BLUE_CIRCLES, blue_circles)
    
    # Serialize adjacency for potential legacy needs or debug
    adj_list = []
    visited_edges = set()
    for u, neighbors in adjacency.items():
        for v in neighbors:
            edge = tuple(sorted((u, v)))
            if edge not in visited_edges:
                visited_edges.add(edge)
                adj_list.append([u, v])
    save_to_redis(KEY_ADJACENCY, adj_list)
    
    return blue_circles, adjacency, relevant_nodes

def create_graph_elements():
    """Step 3: White Lines & Green Circles"""
    logger.info("GraphBuilder: Step 3 - Creating Graph Elements")
    
    blue_circles = load_from_redis(KEY_BLUE_CIRCLES)
    adj_raw = load_from_redis(KEY_ADJACENCY)
    
    relevant_nodes = set()
    coord_to_id = {}
    if blue_circles:
        for bc in blue_circles:
            node = (bc['lat'], bc['lon'])
            relevant_nodes.add(node)
            coord_to_id[node] = bc['id']
    
    adjacency = {}
    if adj_raw:
        for pair in adj_raw:
            u = tuple(pair[0])
            v = tuple(pair[1])
            if u not in adjacency:
                adjacency[u] = set()
            if v not in adjacency:
                adjacency[v] = set()
            adjacency[u].add(v)
            adjacency[v].add(u)

    white_lines = []
    green_circles = []
    visited = set()
    
    relevant_set = set(relevant_nodes)
    
    # Sort for deterministic iteration
    sorted_relevant_nodes = sorted(list(relevant_nodes))
    
    for start_node in sorted_relevant_nodes:
        if start_node not in adjacency:
            continue
        
        # Sort neighbors for deterministic path finding
        neighbors = sorted(list(adjacency[start_node]))
        for neighbor in neighbors:
            edge_key = tuple(sorted((start_node, neighbor)))
            if edge_key in visited:
                continue
            
            path = [start_node, neighbor]
            curr = neighbor
            prev = start_node
            dist = haversine_distance(start_node, neighbor)
            
            while curr not in relevant_set and len(adjacency.get(curr, [])) == 2:
                neighbors = adjacency[curr]
                next_node = next((n for n in neighbors if n != prev), None)
                if next_node:
                    visited.add(tuple(sorted((curr, next_node))))
                    path.append(next_node)
                    dist += haversine_distance(curr, next_node)
                    prev = curr
                    curr = next_node
                else:
                    break
            
            visited.add(edge_key)
            
            if curr in relevant_set and curr != start_node:
                wl = {
                    'id': generate_uid(UIDPrefix.WHITE_LINE),
                    'start': start_node,
                    'end': curr,
                    'start_blue_circle_id': coord_to_id.get(start_node),
                    'end_blue_circle_id': coord_to_id.get(curr),
                    'path': path,
                    'length': dist,
                    'green_count': 0
                }
                
                # Green Circles
                target_spacing = 15.0
                num = max(1, int(round(dist / target_spacing)))
                if num > 1:
                    step = dist / num
                    targets = [step * k for k in range(1, num)]
                    t_idx = 0
                    curr_dist = 0
                    
                    count = 0
                    for i in range(len(path) - 1):
                        p1, p2 = path[i], path[i+1]
                        seg = haversine_distance(p1, p2)
                        while t_idx < len(targets) and (curr_dist + seg) >= targets[t_idx]:
                            rem = targets[t_idx] - curr_dist
                            ratio = rem / seg if seg > 0 else 0
                            nlat = p1[0] + (p2[0] - p1[0]) * ratio
                            nlon = p1[1] + (p2[1] - p1[1]) * ratio
                            green_circles.append({
                                'id': generate_uid(UIDPrefix.GREEN_CIRCLE),
                                'lat': nlat, 'lon': nlon, 
                                'line_id': wl['id']
                            })
                            count += 1
                            t_idx += 1
                        curr_dist += seg
                    wl['green_count'] = count
                
                white_lines.append(wl)
                
    save_to_redis(KEY_WHITE_LINES, white_lines)
    save_to_redis(KEY_GREEN_CIRCLES, green_circles)
    
    return white_lines, green_circles

File: src/services/map/graphBuilder.js
/**
 * Graph Builder - Step 3 of Map Generation
 * Creates white lines and green circles
 */

const config = require('../../config');
const { REDIS_KEYS } = require('../../config/constants');
const { saveToRedis, loadFromRedis } = require('../redis.service');
const { haversineDistance, coordKey, parseCoordKey } = require('../../utils/geometry');

const GREEN_CIRCLE_SPACING = config.game.greenCircleSpacing;

/**
 * Rebuild adjacency map from serialized data
 * @param {Array} adjRaw - Serialized adjacency list
 * @returns {Map} Adjacency map
 */
function rebuildAdjacencyMap(adjRaw) {
  const adjacency = new Map();

  if (!adjRaw) return adjacency;

  for (const pair of adjRaw) {
    const uKey = coordKey(pair[0][0], pair[0][1]);
    const vKey = coordKey(pair[1][0], pair[1][1]);

    if (!adjacency.has(uKey)) adjacency.set(uKey, new Set());
    if (!adjacency.has(vKey)) adjacency.set(vKey, new Set());
    adjacency.get(uKey).add(vKey);
    adjacency.get(vKey).add(uKey);
  }

  return adjacency;
}

/**
 * Get relevant nodes set from blue circles
 * @param {Array} blueCirclesData - Blue circles data
 * @returns {Set} Set of relevant node keys
 */
function getRelevantNodes(blueCirclesData) {
  const relevantNodes = new Set();

  if (!blueCirclesData) return relevantNodes;

  for (const bc of blueCirclesData) {
    relevantNodes.add(coordKey(bc.lat, bc.lon));
  }

  return relevantNodes;
}

/**
 * Trace a path between two intersection nodes
 * @param {string} startNodeKey - Starting node key
 * @param {string} neighborKey - Neighbor node key
 * @param {Map} adjacency - Adjacency map
 * @param {Set} relevantNodes - Set of intersection nodes
 * @param {Set} visited - Set of visited edges
 * @returns {Object|null} { pathCoords, endKey, distance } or null
 */
function tracePath(startNodeKey, neighborKey, adjacency, relevantNodes, visited) {
  const [startLat, startLon] = parseCoordKey(startNodeKey);
  const startNode = [startLat, startLon];

  const [neighborLat, neighborLon] = parseCoordKey(neighborKey);
  const pathCoords = [startNode, [neighborLat, neighborLon]];

  let curr = [neighborLat, neighborLon];
  let currKey = neighborKey;
  let prevKey = startNodeKey;
  let dist = haversineDistance(startNode, curr);

  // Follow chain until we hit another intersection
  while (!relevantNodes.has(currKey) && adjacency.has(currKey) && adjacency.get(currKey).size === 2) {
    const neighborKeys = Array.from(adjacency.get(currKey));
    const nextKey = neighborKeys.find(n => n !== prevKey);
    if (!nextKey) break;

    visited.add([currKey, nextKey].sort().join('|'));

    const [nextLat, nextLon] = parseCoordKey(nextKey);
    const nextNode = [nextLat, nextLon];
    pathCoords.push(nextNode);
    dist += haversineDistance(curr, nextNode);

    prevKey = currKey;
    curr = nextNode;
    currKey = nextKey;
  }

  // Only return if we reached another intersection
  if (relevantNodes.has(currKey) && currKey !== startNodeKey) {
    return { pathCoords, endKey: currKey, endNode: curr, distance: dist };
  }

  return null;
}

/**
 * Create green circles along a white line path
 * @param {Array} pathCoords - Path coordinates
 * @param {number} distance - Total path distance
 * @param {number} lineId - White line ID
 * @returns {Array} Green circles array
 */
function createGreenCirclesForPath(pathCoords, distance, lineId) {
  const greenCircles = [];
  const num = Math.max(1, Math.round(distance / GREEN_CIRCLE_SPACING));

  if (num <= 1) return greenCircles;

  const step = distance / num;
  const targets = [];
  for (let k = 1; k < num; k++) {
    targets.push(step * k);
  }

  let tIdx = 0;
  let currDist = 0;
  let count = 0;

  for (let i = 0; i < pathCoords.length - 1 && tIdx < targets.length; i++) {
    const p1 = pathCoords[i];
    const p2 = pathCoords[i + 1];
    const seg = haversineDistance(p1, p2);

    while (tIdx < targets.length && (currDist + seg) >= targets[tIdx]) {
      const rem = targets[tIdx] - currDist;
      const ratio = seg > 0 ? rem / seg : 0;
      const nlat = p1[0] + (p2[0] - p1[0]) * ratio;
      const nlon = p1[1] + (p2[1] - p1[1]) * ratio;

      greenCircles.push({
        id: `gc_${lineId}_${count}`,
        lat: nlat,
        lon: nlon,
        line_id: lineId
      });

      count++;
      tIdx++;
    }
    currDist += seg;
  }

  return greenCircles;
}

/**
 * Create graph elements (white lines and green circles)
 * @returns {Promise<Object>} { whiteLines, greenCircles }
 */
async function createGraphElements() {
  console.log('Step 3 - Creating Graph Elements');

  const blueCirclesData = await loadFromRedis(REDIS_KEYS.BLUE_CIRCLES);
  const adjRaw = await loadFromRedis(REDIS_KEYS.ADJACENCY);

  const relevantNodes = getRelevantNodes(blueCirclesData);
  const adjacency = rebuildAdjacencyMap(adjRaw);

  const whiteLines = [];
  const greenCircles = [];
  const visited = new Set();

  const sortedNodes = Array.from(relevantNodes).sort();

  for (const startNodeKey of sortedNodes) {
    if (!adjacency.has(startNodeKey)) continue;

    const [startLat, startLon] = parseCoordKey(startNodeKey);
    const startNode = [startLat, startLon];
    const neighbors = Array.from(adjacency.get(startNodeKey)).sort();

    for (const neighborKey of neighbors) {
      const edgeKey = [startNodeKey, neighborKey].sort().join('|');
      if (visited.has(edgeKey)) continue;
      visited.add(edgeKey);

      const pathResult = tracePath(startNodeKey, neighborKey, adjacency, relevantNodes, visited);

      if (pathResult) {
        const lineId = whiteLines.length;

        const wl = {
          id: lineId,
          start: startNode,
          end: pathResult.endNode,
          path: pathResult.pathCoords,
          length: pathResult.distance,
          green_count: 0
        };

        const lineGreenCircles = createGreenCirclesForPath(
          pathResult.pathCoords,
          pathResult.distance,
          lineId
        );

        wl.green_count = lineGreenCircles.length;
        greenCircles.push(...lineGreenCircles);
        whiteLines.push(wl);
      }
    }
  }

  await saveToRedis(REDIS_KEYS.WHITE_LINES, whiteLines);
  await saveToRedis(REDIS_KEYS.GREEN_CIRCLES, greenCircles);

  console.log(`Created ${whiteLines.length} white lines, ${greenCircles.length} green circles`);

  return { whiteLines, greenCircles };
}

module.exports = {
  createGraphElements,
  rebuildAdjacencyMap,
  getRelevantNodes,
  tracePath,
  createGreenCirclesForPath
};
